src/sql/src/func.rs

```diff
@@ -834,37 +605,57 @@ impl From<Vec<ParamType>> for ParamList {
-/// added flexibility.
+/// Describes parameter types.
+///
+/// Parameters with "Compatible" in their name are used in conjunction with
+/// other "Compatible"-type parameters to determine the best common type to cast
+/// arguments to.
+///
+/// "Compatible" parameters contrast with parameters that contain "Any" in their
+/// name, but not "Compatible." These parameters require all other "Any"-type
+/// parameters be of the same type from the perspective of
+/// [`ScalarType::base_eq`].
+///
+/// For more details on polymorphic parameter resolution, see `PolymorphicSolution`.
 pub enum ParamType {
-    /// A pseudotype permitting any type.
+    /// A pseudotype permitting any type. Note that this parameter does not
+    /// enforce the same "Any" constraint as the other "Any"-type parameters.
     Any,
-    /// A polymorphic pseudotype permitting any array type.  For more details,
-    /// see `ParamList::resolve_polymorphic_types`.
+    /// A pseudotype permitting any type, permitting other "Compatibility"-type
+    /// parameters to find the best common type.
+    AnyCompatible,
+    /// An pseudotype permitting any array type, requiring other "Any"-type
+    /// parameters to be of the same type.
     ArrayAny,
-    /// A polymorphic pseudotype permitting all types, with more constraints
-    /// than `Any`, i.e. it is subject to polymorphic constraints. For more
-    /// details, see `ParamList::resolve_polymorphic_types`.
-    ArrayElementAny,
-    /// A polymorphic pseudotype permitting a `ScalarType::List` of any element
-    /// type. For more details, see `ParamList::resolve_polymorphic_types`.
+    /// A pseudotype permitting any array type, permitting other "Compatibility"-type
+    /// parameters to find the best common type.
+    ArrayAnyCompatible,
+    /// An pseudotype permitting any list type, requiring other "Any"-type
+    /// parameters to be of the same type.
     ListAny,
-    /// A polymorphic pseudotype permitting all types, with more constraints
-    /// than `Any`, i.e. it is subject to polymorphic constraints. For more
-    /// details, see `ParamList::resolve_polymorphic_types`.
-    ListElementAny,
-    /// A polymorphic pseudotype with the same behavior as `ListElementAny`,
-    /// except it does not permit either `ScalarType::Array` or
-    /// `ScalarType::List`.
-    NonVecAny,
-    /// A polymorphic pseudotype permitting a `ScalarType::Map` of any non-nested
-    /// value type. For more details, see `ParamList::resolve_polymorphic_types`.
+    /// A pseudotype permitting any list type, permitting other
+    /// "Compatibility"-type parameters to find the best common type.
+    ListAnyCompatible,
+    /// A pseudotype permitting any type, permitting other "Compatibility"-type
+    /// parameters to find the best common type. Additionally, enforces a
+    /// constraint that when used with `ListAnyCompatible`, resolves to that
+    /// argument's element type.
+    ListElementAnyCompatible,
```

Comment:
Is there an easy explanation for why we need this for lists but not arrays?

Reply:
`AnyCompatible` is equivalent to `ArrayElementAnyCompatible`

Reply:
Ah, but then wait: can we not just have `AnyCompatible` serve both purposes?

Reply:
Not if we want to maintain the OID of custom list types on any operation, i.e. if we used `AnyCompatible` and concatenated a an element onto a custom list, you would have to throw away the OID and the return value would always be an anonymous type.

Maybe that's totally fine, but ripping that out is easier than putting it in so figured I'd err on the side of caution.

Reply:
Ohhhh, interesting! Does something go wrong if `guess_common_type` just preserved the the custom list OID in the case of a signature like `(AnyCompatibleList, AnyCompatible)`?

Reply:
Ah the question is then what do you pass to `guess_common_type`. In the case of `AnyCompatible` and `AnyCompatibleArray`:

> Selection of the common type considers the actual types of anycompatible and anycompatiblenonarray inputs, the array element types of anycompatiblearray inputs, the range subtypes of anycompatiblerange inputs, and the multirange subtypes of anycompatiblemultirange inputs.

So you're taking the array's element type. As mentioned, and I'm sure you get, you can't do the same for lists because you'd lose the OID.

The alternative where you would pass the more complex set of types to `guess_common_type` would require an additional pass over the params to determine if you have a `AnyCompatibleList` xor `AnyCompatibleArray` and track that. Then if you encountered any `AnyCompatible`, you'd know to stuff it in as an element of a complex type, and then pass all of those two `guess_common_type`. That seemed fussy to me, though it would accomplish the same thing.